docs/guides/writing_clusters.md

```diff
@@ -1,183 +1,176 @@
-need to be added in these lists.
+---
+
+## Part 1: Cluster Definition (XML)
+
+Clusters are defined based on the Matter specification. The C++ code for them is
+generated from XML definitions located in
+`src/app/zap-templates/zcl/data-model/chip`.
+
+-   **Generate XML:** To create or update a cluster XML, use
+    [Alchemy](https://github.com/project-chip/alchemy) to parse the
+    specification's `asciidoc`. Manual editing of XML is discouraged, as it is
+    error-prone.
+-   **Run Code Generation:** Once the XML is ready, run the code generation
+    script. It's often sufficient to run:
+
+    ```bash
+    ./scripts/run_in_build_env.sh 'scripts/tools/zap_regen_all.py'
+    ```
+
+    For more details, see the
+    [code generation guide](../zap_and_codegen/code_generation.md).
+
+---
+
+## Part 2: C++ Implementation
+
+### File Structure
+
+Create a new directory for your cluster at `src/app/clusters/<cluster-name>/`.
+This directory will house the cluster implementation and its unit tests.
+
+### Recommended Modular Layout
+
+For better testability and maintainability, we recommend splitting the
+implementation into logical components. The
+[Software Diagnostics](https://github.com/project-chip/connectedhomeip/tree/master/src/app/clusters/software-diagnostics-server)
+cluster is a good example of this pattern.
+
+-   **`ClusterLogic`:**
+    -   A type-safe class containing the core business logic of the cluster.
+    -   Manages all attribute storage.
+    -   Should be thoroughly unit-tested.
+-   **`ClusterImplementation`:**
+    -   Implements the `ServerClusterInterface` (often by deriving from
+        `DefaultServerCluster`).
+    -   Acts as a translation layer between the data model (encoders/decoders)
+        and the `ClusterLogic`.
+-   **`ClusterDriver` (or `Delegate`):**
+    -   An optional interface providing callbacks to the application for cluster
+        interactions. We recommend the term `Driver` to avoid confusion with the
+        overloaded term `Delegate`.
+
+### Implementation Details
+
+#### Attribute and Feature Handling
+
+Your implementation must correctly report which attributes and commands are
+available based on the enabled features and optional items.
+
+-   Use a feature map to control elements dependent on features.
+-   Use boolean flags or `BitFlags` for purely optional elements.
+-   Ensure your unit tests cover different combinations of enabled features and
+    optional attributes/commands.
+
+#### Attribute Change Notifications
+
+For subscriptions to work correctly, you must notify the system whenever an
+attribute's value changes.
+
+-   The `Startup` method of your cluster receives a `ServerClusterContext`.
+-   Use the context to call
+    `interactionContext->dataModelChangeListener->MarkDirty(path)`. A
+    `NotifyAttributeChanged` helper exists for paths managed by this cluster.
+    -   For write implementations, you can use `NotifyAttributeChangedIfSuccess`
```

Comment:
Probably some details about how to use the return value from `WriteImpl` to avoid notifying of attributes changes when the write was successful but the value didn't changed?